### ISSUE TITLE
VictoriaLogs returns nothing when limit is < n/2+1

### ISSUE THREAD

**Issue by recht** [VictoriaMetrics/VictoriaMetrics#6785]
### Describe the bug

I'm sending logs to vlogs via the ElasticSearch endpoint from opentelemetry-collector. Metrics indicate that logs are being ingested with no errors, but I don't see anything through vmui. The histogram indicates that there are entries, but nothing shows in the results list, even for a `*` query.

After some experimenting, it turns out that if I remove the `limit` from the query then it returns all entries (around 37000 in my case). Further experimentation shows that if a time range has n entries then the limit must be at least n/2+1 for anything to show up.

### To Reproduce

Count number of entries in a time range and set the limit to 2/n or less. Nothing shows up. With a higher limit, data is returned (but in my case that crashes the browser due to the large data set)

### Version

victoria-logs-20240710-011043-tags-v0.28.0-victorialogs-0-gb173f24041

### Logs

No errors are logged

### Screenshots

```
curl 'https://vlogs/select/logsql/query' -X POST -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0' -H 'Accept: application/stream+json' -H 'Accept-Language: en-US,en;q=0.5' -H 'Content-Type: application/x-www-form-urlencoded;charset=UTF-8'  -H 'Connection: keep-alive' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' -H 'Priority: u=4' -H 'TE: trailers' --data-raw 'query=*&start=2024-08-07T19%3A23%3A12.281Z&end=2024-08-07T19%3A28%3A12.281Z&limit=40000'|wc -l
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 97.0M    0 97.0M  100    87  26.1M     23  0:00:03  0:00:03 --:--:-- 26.1M
37126


curl 'https://vlogs/select/logsql/query' -X POST -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0' -H 'Accept: application/stream+json' -H 'Accept-Language: en-US,en;q=0.5'  -H 'Content-Type: application/x-www-form-urlencoded;charset=UTF-8'  -H 'Connection: keep-alive' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' -H 'Priority: u=4' -H 'TE: trailers' --data-raw 'query=*&start=2024-08-07T19%3A23%3A12.281Z&end=2024-08-07T19%3A28%3A12.281Z&limit=18563'|wc -l
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    87    0     0  100    87      0     62  0:00:01  0:00:01 --:--:--    62
0


curl 'https://vlogs/select/logsql/query' -X POST -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0' -H 'Accept: application/stream+json' -H 'Accept-Language: en-US,en;q=0.5'  -H 'Content-Type: application/x-www-form-urlencoded;charset=UTF-8'  -H 'Connection: keep-alive'  -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' -H 'Priority: u=4' -H 'TE: trailers' --data-raw 'query=*&start=2024-08-07T19%3A23%3A12.281Z&end=2024-08-07T19%3A28%3A12.281Z&limit=18564'|wc -l
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51.8M    0 51.8M  100    87  21.9M     36  0:00:02  0:00:02 --:--:-- 21.9M
18564

```

### Used command-line flags

```
    - --retentionPeriod=1                                                                                                                                               
    - --storageDataPath=/storage                                                                                                                                        
    - --envflag.enable=true                                                                                                                                             
    - --envflag.prefix=VM_                                                                                                                                              
    - --loggerFormat=json      
```

**Comment by Irene-123** [comment]
Hi @jiekun
I went through the issue and looking to contribute here, though need some time for more clarification and understanding
Wanted to know if I can take up this as my first issue here or if you have any suggestions lmk :)
Thanks!

**Comment by jiekun** [comment]
Based on the description:
> set the limit to **2/n** or less. Nothing shows up. ... the limit must be at least **n/2+1** for anything to show up.

It is natural to look for the code that handles the `limit` parameter, especially with special handling for `limit * 2` or `limit / 2`.

So we could find this method:
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/aa9bb995276ec5e6cc1eee5bdb839243da690207/app/vlselect/logsql/logsql.go#L636

This method try to find a suitable rows result whose `limit <= length < 2 * limit` by narrowing down the query time range. And it contains fast path and slow path.

@Irene-123 I would recommand checking this part and find some clues. It would be nice if @recht could provide the dataset you are using.

**Comment by Irene-123** [comment]
HI @recht Would it be possible for you to provide me the dataset?
Also if there any discord channel, lmk!
Thanks

**Comment by Irene-123** [comment]
The issue is with the `d` here, when the limit is very small relative to the initial n/2, the algorithm might incorrectly adjust the start and end times, either overshooting or undershooting the correct time range, leading to no results.

A check we could add is, `start >= end || d == 0` then break the loop

**Comment by jiekun** [comment]
@Irene-123 Here's my test for the algorithm. Assume all logs are generated at `2024-09-08 20:05:11.253308032 +0800 CST`

```
startTime: 2024-09-08 20:02:51.253415603 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:04:06.253415658 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:04:43.753415685 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:02.503415698 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.878415704 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:07.190915701 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:09.534665702 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:10.706540702 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.292478202 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:10.999509452 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.145993827 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.219236014 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.255857107 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.23754656 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.246701833 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST

... // startTime keep increasing

startTime: 2024-09-08 20:05:11.253308032 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.253308033 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
startTime: 2024-09-08 20:05:11.253308032 +0800 CST, endTime: 2024-09-08 20:05:21.253415713 +0800 CST
```

The algorithm could adjust `startTime` step by step until it's equal to the time we mentioned above. I tested some other cases and it seems the algorithm is working correctly. 

> the algorithm might incorrectly adjust the start and end times, either overshooting or undershooting the correct time range

Could you provide some counterexamples?

**Comment by valyala** [comment]
@jiekun , the commit 657988ac3aabad2d2f6021040633a202587b7bf0 contains some optimizations for the algorithm of finding the time range with the given matching `limit=N` logs. The optimization adjusts the upper bound of the time range (`end`) if the current time range has no matching logs.

**Comment by valyala** [comment]
@recht , could you upgrade VictoriaLogs to [v0.29.0-victorialogs](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v0.29.0-victorialogs) and verify whether the issue is still there?

**Comment by recht** [comment]
> @recht , could you upgrade VictoriaLogs to [v0.29.0-victorialogs](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v0.29.0-victorialogs) and verify whether the issue is still there?

I just tried - we had disabled the otel exporter for vmlogs, so nothing had been sent to it for a while. When I enabled it again, I saw logs coming in, but after a couple of minutes it basically went back to the same state: I see counts in the graph, but I don't get any actual log lines out. If I enter a search query other than `*` it just says `No logs volume available
No volume information available for the current queries and time range` (also for something as simple as for example `*a*`)

**Comment by hagen1778** [comment]
> we had disabled the otel exporter for vmlogs

Just want to note, that otel support was added in [v0.29.0-victorialogs](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v0.29.0-victorialogs) and wasn't available before.


>  I see counts in the graph, but I don't get any actual log lines out. If I enter a search query other than * it just says No logs volume available No volume information available for the current queries and time range (also for something as simple as for example *a*)

Could you please provide more information on what exporter do you use for otel and what are its settings? 
Have you followed this guide https://docs.victoriametrics.com/victorialogs/data-ingestion/opentelemetry/ ?

**Comment by recht** [comment]
I was using the elasticsearch exporter pointing to `/insert/elasticsearch/` with `mapping.mode: raw`. I didn't realize there's now otel support, I'll switch to using that.

**Comment by jiekun** [comment]
Thanks for the report. As you mentioned in Slack DM:

> I tried 0.29 and also switched to the otel ingester endpoint, and that seems to work. Previously I was using the elasticsearch ingest endpoint, and that didn't work, also not with 0.29

It could be related to the elasticsearch exporter and VictoriaLogs ingest API.

I tested before without `mapping.mode: raw` and it seems to work. Let me try again.

**Comment by jiekun** [comment]
I was able to reproduce this issue.

By using `mapping.mode: raw`, VictoriaLogs failed to parse the log into https://github.com/VictoriaMetrics/VictoriaMetrics/blob/cc2647d212b4124466e23da69a7e19a76f1f67d5/lib/logstorage/block_header.go#L216

So the `PrefixFilter` could not work correctly.
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/258ccfb9532357af3fdace53ec08cb464b0c5909/lib/logstorage/filter_prefix.go#L57

However, when setting `limit >= n/2+1`, the query goes to the fast path, where `*` filter will be optimize into a `NoopFilter` instead of `PrefixFilter`, so it could retrieve the result:
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/657988ac3aabad2d2f6021040633a202587b7bf0/app/vlselect/logsql/logsql.go#L846

Please note that the query result is actually incorrect: when use any other search condition instead of `*`, it still cannot perform fuzzy match and return the result (because the `PrefixFilter` is not working)

You can reproduce it with my dataset and following command:

[victoria-logs-data.zip](https://github.com/user-attachments/files/17045087/victoria-logs-data.zip)

```
curl 'http://127.0.0.1:9428/select/logsql/query' \
  -H 'Accept: application/stream+json' \
  -H 'Accept-Language: en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7' \
  -H 'AccountID: 0' \
  -H 'ProjectID: 0' \
  --data-raw 'query=*&limit=2&start=2024-09-18T11%3A55%3A35.698Z&end=2024-09-18T12%3A00%3A35.698Z'
```

And the way to avoid it is:
1. Do not use `mapping.mode: raw`.
2. Enhance VictoriaLogs elasticsearch ingest API to handle data in this format.

@valyala @hagen1778

**Comment by valyala** [comment]
@jiekun , thanks for the investigation of the issue!

Could you provide links to docs about `mapping.mode: raw` in the next comment, so future users and/or developers could figure out what is it?

If this is some setting, which is frequently used by standard log shippers, then VictoriaLogs needs to be able to properly parse and store logs received from these log shippers. This is a generic rule of thumb for VictoriaMetrics and VictoriaLogs components - they should properly accept data from popular external clients without the need to configure anything at VictoriaMetrics / VictoriaLogs side. This makes more smooth integration with third-party systems.

**Comment by valyala** [comment]
@jiekun , could you also provide more details to the following statements?

> By using `mapping.mode: raw`, VictoriaLogs failed to parse the log into `columnsHeader` ... so the `PrefixFilter` could not work correctly.

Why VictoriaLogs fails to parse the log in `mapping.mode: raw` format into `columnsHeader`? Does it log parse errors?

> However, when setting `limit >= n/2+1`, the query goes to the fast path, where `*` filter will be optimize into a `NoopFilter` instead of `PrefixFilter`, so it could retrieve the result.

The referenced `q.Optimize()` call at the top of `getLastNQueryResults()` function is executed unconditionally for both cases: `limit < n/2+1` and `limit >= n/2+1`. It is unclear why the `*` filter is replaced with `NoopFilter` only if `limit >= n/2+1`.

**Comment by jiekun** [comment]
Thanks @valyala . Let me do the follow up today.

**Comment by jiekun** [comment]
> Why VictoriaLogs fails to parse the log in mapping.mode: raw format into columnsHeader? Does it log parse errors?

After some tests, I believe the main reason is simply due to the lack of documentation on using the OpenTelemetry Collector with **elasticsearchexporter**. Users did not declare the mapping of the _msg and _time fields in the HTTP Header, which is  necessary.

And as suggested,  rejects the ingested logs without _msg field and with empty _msg field should be a good approach, which  immediately indicated this step and could fix the issue from user side easily.

**Comment by valyala** [comment]
VictoriaLogs should drop ingested logs without `_msg` field or with empty `_msg` field starting from [v0.30.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/) release. See [the comment above](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6785#issuecomment-2376166256) for more details.

Closing the issue as fixed then.
